tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/DataType.java

```diff
@@ -15,49 +15,35 @@
 
 package org.tensorflow;
 
-import java.util.HashMap;
-import java.util.Map;
-
-import org.tensorflow.types.UInt8;
-
-/** Represents the type of elements in a {@link Tensor} as an enum. */
-public enum DataType {
-  /** 32-bit single precision floating point. */
-  FLOAT(1, 4),
-
-  /** 64-bit double precision floating point. */
-  DOUBLE(2, 8),
-
-  /** 32-bit signed integer. */
-  INT32(3, 4),
-
-  /** 8-bit unsigned integer. */
-  UINT8(4, 1),
+import org.tensorflow.internal.c_api.TF_Tensor;
+import org.tensorflow.util.Shape;
+
+/** Represents a type of elements in a {@link Tensor} */
+public final class DataType<T> {
+
+  @FunctionalInterface
+  public interface TensorMapper<T> {
+
+    /**
+     * Maps tensor memory to a data structure for manipulating elements of this type.
+     *
+     * @param nativeTensor pointer to the native tensor
+     * @param shape the shape of the tensor
+     * @return data structure of elements of this type
+     */
+    T apply(TF_Tensor nativeTensor, Shape shape);
+  }
 
   /**
-   * A sequence of bytes.
+   * Creates a new datatype
    *
-   * <p>TensorFlow uses the STRING type for an arbitrary sequence of bytes.
-   */
-  STRING(7, -1),
-
-  /** 64-bit signed integer. */
-  INT64(9, 8),
-
-  /** Boolean. */
-  BOOL(10, 1);
-
-  private final int value;
-  
-  private final int byteSize;
-
-  /**
+   * @param name readable-name for this type
    * @param value must match the corresponding TF_* value in the TensorFlow C API.
    * @param byteSize size of an element of this type, in bytes, -1 if unknown
+   * @param tensorMapper method for mapping tensor memory to elements of this type
    */
-  DataType(int value, int byteSize) {
-    this.value = value;
-    this.byteSize = byteSize;
+  public static <T> DataType<T> create(String name, int value, int byteSize, TensorMapper<T> tensorMapper) {
```

Comment:
Stronger type? `<T extends TType>`.

Reply:
Mmh, I don't remember if that one was as hemorrhagic as with `Tensor`, I'll give it a try and let you know.